### ISSUE TITLE
accessibility options and use of custom styles or xhibit app styles etc

### ISSUE THREAD

**Issue by ronm123** [thexerteproject/xerteonlinetoolkits#818]
Hi all
it was pointed out to me by someone using custom styles created with xhibit app that when either the Invert or Black on Yellow option is selected the footer buttons get messed up. Display of the buttons does indeed get messed up but also the changing to High Contrast or Black on Yellow themes doesn't work correctly either. As you may know we changed the Invert and Black on Yellow options so that they load a separate theme when those options are selected and that's far more reliable in terms of switching everything to white on black or black on yellow but at the moment what that doesn't do is replace custom styles or a custom style sheet. So if for instance the xhibit app theme (or any other custom styles) have background colours set these don't get replaced by the styles in the switched theme.

In many ways this is correct e.g. if someone has added custom styles then they are saying I want these styles however should we change things so that the accessibility options when selected overwrite custom styles too? @torinfo any thoughts on this?

**Comment by james-roscoe** [comment]
Thanks for flagging this up Ron. Xhibit already has some limited functionality to ensure the invert/black-on-yellow functionality is preserved for text ([via this commit](https://github.com/james-roscoe/XhibitApp/commit/099ee10e0a946d95e0e2a6c9255ebad901747913)) but we haven't updated that since we added custom FontAwesome icons in. I'm happy to update Xhibit to ensure the icons are still visible when these options are selected, though as you say it would make sense for there to be a solution in Xerte itself which ensures the accessibility functions will still work for any custom theme.

I'd be a little wary of Tom's suggestion of turning off custom stylesheets entirely though. I appreciate this would be the simplest way of handling it, but as mentioned this would break other functionality (quite severely on many of my previous projects, actually). I think the accessibility options should override the text/background colours used in custom themes, but CSS relating to other layout areas needn't be effected, imho...

**Comment by ronm123** [comment]
@james-roscoe the problem is the accessibility options haven't really worked properly since the days when there was only one theme. With all the themes and colour combinations we've had since the only reliable way to display via high contrast or black on yellow and affect most of the content as well as the interface was to switch themes altogether to a high contrast theme or a black on yellow theme. I think your commit pre-dated that change so no longer works anyway. At the moment as soon as there is custom styles whether xhibit or from elsewhere that currently breaks depending on what styles have been added. Obviously if someone know what they are doing and only uses styles not likely to affect the accessibility options it's fine but we have those accessibility options to make it easy for end users to choose one of those themes.

**Comment by JoelReed** [comment]
Please note that we've now addressed the Invert and BOY accessibility options for Xhibit App and these fixes were handled in a [recent commit](https://github.com/james-roscoe/XhibitApp/pull/24), along with a further fix for the [glossary tool](https://github.com/james-roscoe/XhibitApp/pull/25). Users can now toggle between their Xhibit stylesheet and the core accessibility colour options defined by Xerte. This has been built into the theme export over at [xhibitapp.com](https://xhibitapp.com).

**Comment by ronm123** [comment]
@JoelReed I was surprised by this so just tested and I may be misunderstanding what you have said now works but according to my testing no change to xhibit app css will fix this at the moment because the switching to invert or black on yellow loads those entirely separate themes but doesn't do anything with the invert and boy classes. So if an author using xhibit has set header, footer and background colours etc those colours still appear when an end user chooses invert or black on yellow. e.g. those colours set with xhibit css do not change.

**Comment by JoelReed** [comment]
@ronm123 - Sorry for any confusion. If you open this [example resource](https://www.nottingham.ac.uk/toolkits/play_20324), which features the latest Xhibit stylesheet, and try toggling between the accessibility options ('Invert'/'BOY') and the Xhibit theme (i.e. 'Off'), you will notice that the correct accessibility colour scheme is now showing on the 'Invert' and 'BOY' options for all elements. These changes have been made centrally on the Xhibit App theme targeting the accessibility classes and builds [on a fix](https://github.com/james-roscoe/XhibitApp/commit/099ee10e0a946d95e0e2a6c9255ebad901747913) that James did a while ago.

**Comment by ronm123** [comment]
@JoelReed ok I did wonder whether this might be the case - I'm guessing that you aren't testing with the latest develop code or Xerte X? I'm not sure the UoN install is up-to-date.

Here's a link to a quick test/demo of the problem https://xot.xerte.org.uk/play.php?template_id=141 
If you read through this issue again you'll see that this isn't an issue you can fix via xhibit but likewise it's a challenge because of what @james-roscoe has said too. The problem is if someone chooses black on yellow or effectively white on black for invert then as much as possible should change to that combination. The orginal way of dealing with this worked when there weren't so many themes and there wasn't all the editor styles etc but has long since not really been effective. That's why we changed it so a black on yellow and high contrast theme are loaded instead but at the moment any custom css isn't changing that. 

@james-roscoe has said that loading the separate theme over the custom css would break your existing projects. I think we need to unpick exactly how and why that is the case because the priority should be if a user chooses black on yellow as far as possible they get black on yellow.

**Comment by JoelReed** [comment]
Yes. The fixes are applicable to Xerte versions 3.5 and earlier. Nottingham currently runs 3.5. Back to the drawing board for Xerte X, as discussed above.

**Comment by FayCross** [comment]
@ronm123 I've done some work on the accessibility themes recently and have stumbled across this old issue. It's still the case that if a custom style sheet from xhibit is used then the accessibility theme buttons display is wrong. It would be easy to change it so that custom style sheets are not used when an alternative theme is being used - only when the main theme is set to xhibit. What do you think? I'm reluctant to always remove the custom stylesheet when an accessibility theme is in use but perhaps limiting it to only when xhibit theme is in use would be a good compromise?

**Comment by ronm123** [comment]
@FayCross I'm not sure what the best solution is here but from memory wasn't there also a fix in xhibit and/or Xerte tp resolve the button display issues and could it also be that the creation date of custom xhibit stylesheets could also be a factor? e.g. does this vary by an older vs newer xhibit style sheet? It's always concerned me that xhibit stylesheets become a point in time version divorced from later changes in Xerte and probably little attention to that by the creators of those stylesheets.

**Comment by FayCross** [comment]
I've just tried with a newly created stylesheet from the xhibit app and it's still a problem. The buttons are the obvious thing but you can see the black/yellow theme is mostly overridden:
![Image](https://github.com/user-attachments/assets/612e2884-c9ae-4479-803f-31472e5b64e7)

**Comment by ronm123** [comment]
@FayCross perhaps there's a short term and long term solution here:

Short term fix for as it is now.
Longer term fix once @torinfo has migrated the xhibitapp and we add a solution in that too?

I'm certainly aware of projects where the xhibit theme has been selected but then custom styles that haven't been created with xhibit added, but likewise projects where other themes are selected but xhibit styles added as well as projects where the xhibit theme and then xhibit styles have been added. So I guess there are three variations as well as any difference with the time of creation.

**Comment by FayCross** [comment]
@ronm123 @torinfo The problems with footer button icons when using an xhibit stylesheet & an accessibility theme is in use has been raised in the testing that the documentation team have been doing. I have made a change so that if the xhibit theme is in use and an accessibility theme is chosen by the end-user, any custom stylesheet uploaded to the project is disabled. This fixes the problem.

We should probably think about whether we always disable the custom stylesheet (& maybe custom styles too) when accessibility theme is used, regardless of whether the project uses the xhibit theme or not. I think it would probably be a good thing to do this as I imagine most custom stylesheets will be used for purely visual things like colours that may clash with the accessibility themes. What do you think?

**Comment by FayCross** [comment]
Closing as we've discussed and decided to only disable stylesheets if xhibit theme is in use
